packages/client/src/objectSet/ObjectSet.test.ts

```diff
@@ -436,6 +437,260 @@ describe("ObjectSet", () => {
+          // @ts-expect-error
+          base.pivotTo("lead").selectProperty("notAProperty");
+
+          return base.pivotTo("lead").selectProperty("employeeStatus");
+        },
+      });
+    });
+
+    it("propagates derived property type to future object set operations with correct types", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").aggregate("employeeId:sum"),
+        // @ts-expect-error
+      }).where({ "notAProperty": { "$eq": 3 } });
+
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").aggregate("employeeId:sum"),
+      }).where({ "derivedPropertyName": { "$eq": 3 } });
+
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").aggregate("employeeId:collectList"),
+      }).where({ "derivedPropertyName": { "$isNull": false } })
+        // @ts-expect-error
+        .where({ "derivedPropertyName": { "$eq": [1, 2] } });
+
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").aggregate("employeeId:collectSet"),
+      }).where({ "derivedPropertyName": { "$isNull": false } })
+        // @ts-expect-error
+        .where({ "derivedPropertyName": { "$eq": [1, 2] } });
+
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").selectProperty("employeeId"),
+      }).where({ "derivedPropertyName": { "$eq": 3 } });
+
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").selectProperty("startDate"),
+      }).where({ "derivedPropertyName": { "$eq": "datetimeFilter" } });
+    });
+
+    it("correctly types multiple property definitions in one clause", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").aggregate("employeeId:sum"),
+        "derivedPropertyName2": (base) =>
+          base.pivotTo("lead").selectProperty("fullName"),
+      }).where({ "derivedPropertyName": { "$eq": 3 } })
+        .where({ "derivedPropertyName2": { "$eq": "name" } });
+    });
+
+    it("ensures other properties are consistently typed", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").selectProperty("employeeId"),
+      }).where({ "fullName": { "$eq": "A" } });
+
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").selectProperty("employeeId"),
+      }).where({ "employeeId": { "$eq": 2 } });
+    });
+
+    it("allows fetching derived properties with correctly typed Osdk.Instance types", async () => {
+      const objectWithRdp = await client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          base.pivotTo("lead").selectProperty("employeeId"),
```

Comment:
Can we add a test to verify things don't break if we select a property that has all undefined values?

Reply:
What do you mean?

Reply:
Basically if a prop returns an undefined value